libraries/exoplayer_rtsp/src/main/java/androidx/media3/exoplayer/rtsp/reader/RtpVP9Reader.java

```diff
@@ -0,0 +1,230 @@
+        + Util.scaleLargeTimestamp(
+            (rtpTimestamp - firstReceivedRtpTimestamp),
+            /* multiplier= */ C.MICROS_PER_SECOND,
+            /* divisor= */ MEDIA_CLOCK_FREQUENCY);
+  }
+
+  private static boolean parseVP9Descriptor(ParsableByteArray payload, int packetSequenceNumber)
+      throws ParserException {
+    // VP9 Payload Descriptor, Section 4.2
+    //         0 1 2 3 4 5 6 7
+    //        +-+-+-+-+-+-+-+-+
+    //        |I|P|L|F|B|E|V|Z| (REQUIRED)
+    //        +-+-+-+-+-+-+-+-+
+    //   I:   |M| PICTURE ID  | (RECOMMENDED)
+    //        +-+-+-+-+-+-+-+-+
+    //   M:   | EXTENDED PID  | (RECOMMENDED)
+    //        +-+-+-+-+-+-+-+-+
+    //   L:   | TID |U| SID |D| (Conditionally RECOMMENDED)
+    //        +-+-+-+-+-+-+-+-+
+    //        |   TL0PICIDX   | (Conditionally REQUIRED)
+    //        +-+-+-+-+-+-+-+-+
+    //   V:   | SS            |
+    //        | ..            |
+    //        +-+-+-+-+-+-+-+-+
+
+    int header = payload.readUnsignedByte();
+    if (!gotFirstPacketOfVP9Frame) {
+      // For start of VP9 partition B=1 as per VP9 RFC Section 4.2.
+      if ((header & 0x08) == 0) {
+        Log.w(
+            TAG,
+            Util.formatInvariant(
+                "First payload octet of the RTP packet is not the beginning of a new VP9 partition,"
+                    + " Dropping current packet." + header));
+        return false;
+      }
+      gotFirstPacketOfVP9Frame = true;
+    } else {
+      // Check that this packet is in the sequence of the previous packet.
+      int expectedSequenceNumber = RtpPacket.getNextSequenceNumber(previousSequenceNumber);
+      if (packetSequenceNumber != expectedSequenceNumber) {
+        Log.w(
+            TAG,
+            Util.formatInvariant(
+                "Received RTP packet with unexpected sequence number. Expected: %d; received: %d."
+                    + " Dropping packet.",
+                expectedSequenceNumber, packetSequenceNumber));
+        return false;
+      }
+    }
+
+    // Check I optional header present.
+    if ((header & 0x80) != 0) {
+      int optionalHeader = payload.readUnsignedByte();
+      // Check M for 15 bits PictureID.
+      if ((optionalHeader & 0x80) != 0) {
+        if (payload.bytesLeft() < 1) {
+          return false;
+        }
+      }
+    }
+
+    // Flexible-mode not implemented.
+    checkArgument((header & 0x10) == 0, "VP9 flexible mode unsupported");
+
+    // Check L optional header present.
+    if ((header & 0x20) != 0) {
+      payload.skipBytes(1);
+      if (payload.bytesLeft() < 1) {
+        return false;
+      }
+      // Check TL0PICIDX header present (non-flexible mode).
+      if ((header & 0x10) == 0) {
+        payload.skipBytes(1);
+      }
+    }
+
+    // Check V optional header present, Refer Section 4.2.1.
+    if ((header & 0x02) != 0) {
+      int scalabilityStr = payload.readUnsignedByte();
+      int numSpatialLayers = (scalabilityStr & 0xe0) >> 5;
```

Comment:
`(scalabilityStr >> 5) & 0x7`

If you do `& 0xe0` and then shift, there could be sign extension. Basically `0xF0000000 >> 1` is `0xF8000000` instead of `0x78000000`